neqo-transport/src/crypto.rs

```diff
@@ -1358,6 +1360,80 @@ pub enum CryptoStreams {
     },
 }
 
+/// [Fisher–Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#Modern_method)
+/// Modified to make sure no element stays in place if `a` has more than one element.
+///
+/// TODO: We could use <https://docs.rs/rand/latest/rand/seq/trait.SliceRandom.html#tymethod.shuffle>
+/// instead, but I assume we want to only use the NSS RNG in neqo?
+fn shuffle<T>(a: &mut [T]) {
+    // To shuffle an array a of n elements (indices 0..n-1)
+    let n: usize = a.len();
+    if n < 2 {
+        return;
+    }
+    for i in 0..(n - 1) {
+        // j ← random integer such that i ≤ j < n
+        let j = loop {
+            let j = (random::<8>()[0] as usize) % (n - i) + i;
```

Comment:
Calling `random::<8>` samples 8 bytes, but you are only using one.

This is going to more efficient if you sample once.  I suggest that - with the removal of the rejection sampling loop - you do this:

```rust
let mut randomness = [0; 32]; // set the value to what you prefer
let count = min(randomness.len(), n - 1);
randomize(&mut randomness[..count]);
for (i, r) in randomness.iter().enumerate().take(count) {
   let j = usize::from(r) % (n - i - 1) + i + 1;
   a.swap(i, j);
}
```

This will also limit the number of cuts that you swap over.  That makes it less random overall, but that's fine.

Reply:
> Calling random::<8> samples 8 bytes, but you are only using one.

Good catch. I meant to use them all.